### PR TITLE
Direction/gradient clipping

### PR DESCRIPTION
**Pull request by MarcCote**
This PR first introduce the concept of gradient modifiers: `loss.append_gradient_modifier(...)`. These modifiers are exactly like direction modifiers but are applied on the gradient. This was needed since gradient modifiers will be applied **before** computing moving average or momentum found in optimizers like Adagrad, Adadelta, RMSprop, Adam etc.

This PR also add a new direction modifier, `DirectionClipping`. Note that it can also be used as a gradient modifier.

### CODE REVIEW

File: smartlearner/interfaces/loss.py [hunk 1]
```diff
@@ -83,14 +88,29 @@ def updates(self):
...
-        return self.gradients
+        self.orig_gradients = OrderedDict(zip(self.model.parameters, gparams))
+
+        # Apply gradients modifiers and gather updates from these modifiers.
+        gradients = self.orig_gradients.copy()
+        self._gradient_modifiers_updates = self._apply_modifiers(self._gradient_modifiers, gradients)
+        return gradients
+
+    def _apply_modifiers(self, list_modifiers, objects_to_modify):
+        # TODO: this is a copy-paste of the method in the class Optimizer.
```
Comment (MarcCote) [inline smartlearner/interfaces/loss.py:106]: Maybe we could move that function inside the `Modifier` classes?

**Comment by MarcCote** [comment]
@ASalvail, @ppoulin91 : either one of you could have a look at this I think we'll all need it soon.

**Comment by ppoulin91** [comment]
Looks good to me

**Comment by ASalvail** [comment]
No.
Just no.

We won't introduce yet another modifier. Instead of making a new one, we should instead unify them all.
Smartlearner is getting ridiculously rigid to do anything in, and I don't think this will help any.

Before I can consider this PR, let's rather brainstorm on how to get rid of all modifiers.

**Comment by MarcCote** [comment]
I see your point and this PR is a first step in that direction. There is no `GradientModifier` class, only a method in the `Loss` class that allows _modifying_ the gradient (e.g. do to gradient clipping). That said function takes as input any existing `DirectionModifier` object.

What do you think of removing the class `ParamModifier` and renaming `DirectionModifier` to just `Modifier`. With that, if one desires to modify: 
- the gradient, then he can append a modifier to the loss via `loss.append_gradient_modifier`
- the direction used to update the params, then he can append a modifier to the optimizer via `optimizer.append_direction_modifier`
- the params after the update, then he can append a modifier to the optimizer via `optimizer.append_param_modifier`.

I agree the last one might not especially useful right now (maybe add a constraint of positivity on the params) but as you said "we want Smartlearner to be as flexible as possible".

**Comment by ASalvail** [comment]
That actually sounds a lot better than having all these subclasses. It remains flexible, but simpler than it is now. My only concern is probably the growth of latent complexity.

The more I think about the problem, the more I see a functional approach à la `numpy` to be a possible solution. This way, we don't hide away the complexity, which in turn permits to be more flexible but still allows for encapsulation of often used tricks. However, that'd be a major rewrite...

**Pull request merged**